### ISSUE TITLE
Cannot add a position_in_category using webservices

### ISSUE THREAD

**Issue by lewis2413** [PrestaShop/PrestaShop#14903]
###Describe the bug
I´m trying to change the position of a product into category using the prestashop api but don´t working for me since that webservice generates a ramdon position skiping that i have done.

To reproduce this error I use the postman tool and the xml product structure

### Expected behavior

If we have an empty category and we set position_in_category to 1, the product is added into the category.

### Steps to reproduce

1 . have empty category
2 . call API to create a product into that category passing position_in_category equals to 1.

#### Acceptance Test
Unskip tests 
* https://github.com/PrestaShop/PrestaShop/blob/d6bd4586da1d519e3ccb6ce2dd2420b9a755b5b2/tests/UI/campaigns/functional/WS/02_productsCRUD.ts#L442-L443
* https://github.com/PrestaShop/PrestaShop/blob/d6bd4586da1d519e3ccb6ce2dd2420b9a755b5b2/tests/UI/campaigns/functional/WS/02_productsCRUD.ts#L1123-L1124

### Additional information
PrestaShop version: 1.7.5.0 & 1.7.8.8
PHP version: 7.0.33

**Comment by khouloudbelguith** [comment]
Hi @lewis2413,

I manage to reproduce the issue with PS1.7.6.0 & PS1.7.5.2.
I attached a screen record
https://drive.google.com/file/d/1kTDiNC19Xk5kd0m-vhijwK9axB8laymI/view
I used this script to update the position of a product
<html><head><title>CRUD Tutorial - Create example</title></head><body>

```
<?php
/**
 * Created by PhpStorm.
 * User: khouloud.belguith
 * Date: 30/09/17
 * Time: 11:58
 */
define('DEBUG', true);
ini_set('display_errors','on');
define('PS_SHOP_PATH', 'http://shop.com');
define('PS_WS_AUTH_KEY', 'Key');
require_once('PSWebServiceLibrary.php');
$webService= new PrestaShopWebservice(PS_SHOP_PATH, PS_WS_AUTH_KEY, DEBUG);
$id =  23; 
$category_id = 3;
$xml = $webService->get(array('resource' => 'products', 'id' => $id));
unset($xml->product->manufacturer_name);
unset($xml->product->quantity);

$product = $xml->children()->children();
$product->name->language[0][0] = "Produit webservice";

$product->description->language[0][0] = "Description produit webservice";

$product->price = 1.121;
$product->position_in_category = 1;
//I don't update XML, i just resend it here
$opt['putXml'] = $xml->asXML();                  
$opt['id'] = $id;
$opt['category_id'] = $category_id;
$opt['resource'] = 'products';

$xml = $webService->edit($opt);

echo "DONE";
```

I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

**Comment by lewis2413** [comment]
Hi,

It´s imposible to change the position of a product inside a category using
the originals methods from the prestashop api, when i launch a  request
internaly i dont what it does, but it changes the custom position and it
always add +1 to my position.
*--*

*Saludos Cordiales*


*Ing. Jose Luis Idrogo*
Movil: +34 678 878385
Email: joseluis.idrogo@edisa.com
Web Site: http://www.edisa.com/
* <http://www.edisa.com/>*
*MADRID - BARCELONA - VIGO - OVIEDO - OURENSE - LAS PALMAS - MEXICO -
COLOMBIA - ECUADOR - REP. DOMINICANA - PANAMÁ*
*------------------------------CONFIDENCIALIDAD:* Este mensaje se dirige
exclusivamente a su destinatario y puede contener información privilegiada
o confidencial. Si no es vd. el destinatario indicado, queda notificado de
que la utilización, divulgación y/o copia sin autorización está prohibida
en virtud de la legislación vigente. Si ha recibido este mensaje por error,
le rogamos que nos lo comunique inmediatamente por esta misma vía y proceda
a su destrucción.

Antes de imprimir este mensaje, asegúrate de que es necesario. Proteger el
medio ambiente está también en tu mano.

**Comment by lewis2413** [comment]
Hi good afternoon,

@khouloudbelguith Do you have any feedback about this problem? i´m contacting to you because we need some solution about this bug..

I'll hope your comments

thanks advance

**Comment by khouloudbelguith** [comment]
Hi @lewis2413,

Sorry, this issue is not yet.
There are some major issues to solve before this one.
But PrestaShop is an open source project, so it can be solved before if someone submits a pull request to solve it.
Thanks!

**Comment by Jeydh** [comment]
I'm not sure it's related to this problem but the function setWsCategories() in classes/Product.php looks to work in a wrong way to me.
No matter what you send in position_in_category, the datas in category_product table (sorted by product id) is deleted and recreated as if it was sorted by the category id.
Feel free to check that by yourself, it's only my observations, maybe I got something wrong.

I'm not a pro in PrestaShop code so I won't dare to submit a pull request but I hope those infos will help to fix this issue.

**Comment by TomLorenzi** [comment]
I've made a PR normally handling this bug : https://github.com/PrestaShop/PrestaShop/pull/31342

**Comment by afdev82** [comment]
I am getting this bug too... The category is not empty, and I can't set the `position_in_category` for the new product, because I get this error:

> Error on creating product: This call to PrestaShop Web Services failed and returned an HTTP status of 500. That means: Internal Server Error - (Code 135): Non puoi impostare una posizione maggiore del numero complessivo dei prodotti nella categoria, che comincia da 1.

Position in category is 4 and in the category I have already 3 products.
I have found why: the PR #16697 changed [the condition](https://github.com/PrestaShop/PrestaShop/pull/16697/files#diff-e94a594ba740485c7a4882b333984d3932a2f99c0d6d0005620745087cce7a10R6146) in the Product.php class.
For a new product that cannot work, because its position will be always bigger of the current count.
I will try to fix it myself and see, but I am new to Prestashop.

The problem is that I get a new product without any `position_in_category` set and the result is that in the cart I can't update the quantity of that product, I get an error (in the developer console of the browser) that the product is not available. In the front end the quantity returns back to the previous value.
If I save the product in the backed, I can update the quantity in the cart and I notice that `position_in_category` has been set.

**Comment by TomLorenzi** [comment]
@afdev82 If you want to spend some times you can juste use my previous PR https://github.com/PrestaShop/PrestaShop/pull/31342 and fix the bug that was reported by the QA Team, the logic was already validated.

**Comment by afdev82** [comment]
@TomLorenzi Yes, I have already seen that PR, but in my opinion the real problem in that function is that it cannot update a row that doesn't exist in the DB (in the case of a new product created with the Webservice: https://github.com/PrestaShop/PrestaShop/pull/31342/files#diff-e94a594ba740485c7a4882b333984d3932a2f99c0d6d0005620745087cce7a10R7195

I don't think I could find a general solution that could work, because I don't know so much Prestashop.
As I can see the problem with a completely new product is that in the `setWsPositionInCategory` function we don't have yet the product ID, then I can't create the row for the position in the `category_product` table.

Maybe someone with more experience could help here, for the moment I am just creating the row in my class, where I know (after the product creation) the product ID.

**Comment by boo-code** [comment]
Still reproduces on 9.2.x. Draft PR: #42473.

The cause is the order inside `WebserviceRequestCore::saveEntityFromXml()`: field setters run at line 1588, the object is saved at 1636, associations are applied at 1656. So `setWsPositionInCategory()` runs while the product has no id and no row in the category. It places nothing, and the association that follows appends at `MAX(position) + 1`. Measured against a category of 5 products, asking for 3 returns the product at 6.

It also does not leave the category alone on the way out. Not finding the product in the list it still ran the renumbering loop: a category reading `3@1 4@3 5@4 12@5 13@6 14@7 15@8` came back as `3@1 4@2 5@3 12@4 13@5 14@6 15@7` from a call that placed nothing.

Two more defects on the same path, both included:

- The bounds check refused the position after the last one, although a product joining a category adds a slot to it. That is what #31342 by @TomLorenzi fixed; it was approved twice and closed in December 2023 for lack of time, and the PR credits it.
- `setWsCategories()` calls `deleteCategories()` with `$clean_positions` defaulting to false, so removing a category from a product's associations vacates a position without renumbering. After that the stored position no longer indexes the ordered list: with a category reading `1, 3, 4, 5, 6`, asking the product at 6 to move to 2 moved the product at 3 instead.

The acceptance criteria in the description are unskipped in the PR. The create payload asks for position 3 rather than the 1 in the commented-out line: 1 passes with or without the fix, because a product the setter cannot place sorts first anyway and lands at 1 by accident.
